Python/gc_free_threading.c

```diff
@@ -1,9 +1,1660 @@
+                 - do less collections
+               This heuristic is about the latter strategy.
+
+               In addition to the various configurable thresholds, we only trigger a
+               full collection if the ratio
+
+                long_lived_pending / long_lived_total
+
+               is above a given value (hardwired to 25%).
+
+               The reason is that, while "non-full" collections (i.e., collections of
+               the young and middle generations) will always examine roughly the same
+               number of objects -- determined by the aforementioned thresholds --,
+               the cost of a full collection is proportional to the total number of
+               long-lived objects, which is virtually unbounded.
+
+               Indeed, it has been remarked that doing a full collection every
+               <constant number> of object creations entails a dramatic performance
+               degradation in workloads which consist in creating and storing lots of
+               long-lived objects (e.g. building a large list of GC-tracked objects would
+               show quadratic performance, instead of linear as expected: see issue #4074).
+
+               Using the above ratio, instead, yields amortized linear performance in
+               the total number of objects (the effect of which can be summarized
+               thusly: "each full garbage collection is more and more costly as the
+               number of objects grows, but we do fewer and fewer of them").
+
+               This heuristic was suggested by Martin von Löwis on python-dev in
+               June 2008. His original analysis and proposal can be found at:
+               http://mail.python.org/pipermail/python-dev/2008-June/080579.html
+            */
+            if (i == NUM_GENERATIONS - 1
+                && gcstate->long_lived_pending < gcstate->long_lived_total / 4)
+            {
+                continue;
+            }
+            return i;
+        }
+    }
+    return -1;
+}
+
+static void
+cleanup_worklist(struct worklist *worklist)
+{
+    PyObject *op;
+    while ((op = worklist_pop(worklist)) != NULL) {
+        gc_restore_tid(op);
+        gc_clear_unreachable(op);
+        Py_DECREF(op);
+    }
+}
+
+static void
+gc_collect_internal(PyInterpreterState *interp, struct collection_state *state)
+{
+    _PyEval_StopTheWorld(interp);
+    // Find unreachable objects
+    int err = deduce_unreachable_heap(interp, state);
+    if (err < 0) {
+        _PyEval_StartTheWorld(interp);
+        goto error;
+    }
+
+    // Print debugging information.
+    if (interp->gc.debug & _PyGC_DEBUG_COLLECTABLE) {
+        PyObject *op;
+        WORKSTACK_FOR_EACH(&state->unreachable, op) {
+            debug_cycle("collectable", op);
+        }
+    }
+
+    // Clear weakrefs and enqueue callbacks (but do not call them).
+    clear_weakrefs(state);
+    _PyEval_StartTheWorld(interp);
```

Comment:
So I think if I understand this correctly what's going to happen with the thread ID and any reference count changes that may happen at this point, but the assumption is that the thread ID is never going to actually clash with an object reference in the work list.  And then anything which does have a reference count operation will just become merged when we restore it?

Reply:
Yeah, the key assumption is that `_PyThread_Id()` is itself a pointer to a distinct object so never conflicts with an object in the worklists. The implementation of `_PyThread_Id()` uses the address of the platform's thread control block or equivalent.

On the GC side, we abuse `ob_tid` for two purposes:

* `gc_refs`, which can have arbitrary values (might conflict!), but this is only happens while other threads are paused, so nobody sees it.
* `worklist` pointers, which must not conflict with `_PyThread_Id()`

Some worklists are only used while other threads are paused (e.g., in `_PyGC_GetObjects()`). Other worklists are created when threads are paused, but still used while other threads may be running (e.g., `unreachable`, `wrcb_to_call`).

If the worklist continues to be used while other threads are running, then there are two other important considerations:

* The worklist holds a strong reference to the object so that it can't be freed while in the worklist.
* The object's reference count fields are merged before adding it to the worklist. This avoids some other thread trying to merge the refcount fields, which would be messy. (It also makes the deallocation happen more quickly when we get around to calling `tp_clear`).